pkg/scheduler/framework/plugins/volumebinding/volume_binding.go

```diff
@@ -42,15 +42,56 @@ func (pl *VolumeBinding) Name() string {
 	return Name
 }
 
+func podHasPVCs(pod *v1.Pod) bool {
+	for _, vol := range pod.Spec.Volumes {
+		if vol.PersistentVolumeClaim != nil {
+			return true
+		}
+	}
+	return false
+}
+
 // Filter invoked at the filter extension point.
+// It evaluates if a pod can fit due to the volumes it requests,
+// for both bound and unbound PVCs.
+//
+// For PVCs that are bound, then it checks that the corresponding PV's node affinity is
+// satisfied by the given node.
+//
+// For PVCs that are unbound, it tries to find available PVs that can satisfy the PVC requirements
+// and that the PV node affinity is satisfied by the given node.
+//
+// The predicate returns true if all bound PVCs have compatible PVs with the node, and if all unbound
+// PVCs can be matched with an available and node-compatible PV.
 func (pl *VolumeBinding) Filter(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *framework.Status {
-	_, reasons, err := pl.predicate(pod, nil, nodeInfo)
-	return migration.PredicateResultToFrameworkStatus(reasons, err)
+	node := nodeInfo.Node()
+	// If pod does not request any PVC, we don't need to do anything.
+	if !podHasPVCs(pod) {
+		return nil
+	}
+
+	unboundSatisfied, boundSatisfied, err := pl.binder.Binder.FindPodVolumes(pod, node)
+
+	failReasons := []predicates.PredicateFailureReason{}
+	if !boundSatisfied {
+		failReasons = append(failReasons, predicates.ErrVolumeNodeConflict)
+	}
+
+	if !unboundSatisfied {
+		failReasons = append(failReasons, predicates.ErrVolumeBindConflict)
+	}
+
+	if len(failReasons) <= 0 {
+		return nil
+	}
+
+	// All volumes bound or matching PVs found for all unbound PVCs
+	return migration.PredicateResultToFrameworkStatus(failReasons, err)
```

Comment:
Please replace all the above with the following so that we start cutting dependency on migration package:

```
if !boundSatisfied || !unboundSatisfied {
  status := framework.NewStatus(UnschedulableAndUnresolvable)
  if !boundSatisfied {
    status.AppendReason(predicates.ErrVolumeNodeConflict.GetReason())
  } 
  if !unboundSatisfied {
    status.AppendReason(predicates.ErrVolumeBindConflict.GetReason())
  } 
  return status
}
return nil
```